src/sentry/auth/providers/saml2.py

```diff
@@ -1,65 +1,134 @@
 from sentry.utils.http import absolute_uri
+from sentry.utils.auth import login, get_login_redirect
+
+
+def get_provider(organization_slug):
+    try:
+        organization = Organization.objects.get(
+            slug=organization_slug
+        )
+    except Organization.DoesNotExist:
+        return HttpResponseRedirect(reverse('sentry-login'))
+
+    if organization.status != OrganizationStatus.VISIBLE:
+        return HttpResponseRedirect(reverse('sentry-login'))
+
+    try:
+        auth_provider = AuthProvider.objects.get(
+            organization=organization
+        )
+        return auth_provider.get_provider()
+    except AuthProvider.DoesNotExist:
+        return HttpResponseRedirect(reverse('sentry-login'))
 
 
 class SAML2LoginView(AuthView):
     def dispatch(self, request, helper):
-        if 'SAMLResponse' in request.POST.keys():
-            return helper.next_step()
-        saml_config = helper.provider.build_saml_config(helper.organization.slug)
-        auth = helper.provider.build_auth(request, saml_config)
+        provider = helper.provider
+        saml_config = provider.build_saml_config(helper.organization.slug)
+        auth = provider.build_auth(request, saml_config)
         return self.redirect(auth.login())
 
 
 class SAML2ACSView(AuthView):
-    def dispatch(self, request, helper):
-        saml_config = helper.provider.build_saml_config(helper.organization.slug)
-        auth = helper.provider.build_auth(request, saml_config)
+    @method_decorator(csrf_exempt)
+    def dispatch(self, request, organization_slug):
+        provider = get_provider(organization_slug)
+        organization = Organization.objects.get(slug=organization_slug)
+        saml_config = provider.build_saml_config(organization_slug)
+
+        auth = provider.build_auth(request, saml_config)
         auth.process_response()
         errors = auth.get_errors()
         if errors:
             error_reason = auth.get_last_error_reason()
             raise IdentityNotValid(error_reason)
 
-        data = {
-            'attributes': auth.get_attributes(),
-            'nameid': auth.get_nameid(),
-            'session_index': auth.get_session_index()
-        }
+        attributes = auth.get_attributes()
+        nameid = auth.get_nameid()
 
-        helper.bind_state('data', data)
-        return helper.next_step()
-        #return HttpResponse("ACS endpoint")
+        email = self.retrieve_email(attributes, nameid, provider.config)
 
+        # Filter users based on the emails provided in the commits
+        user_emails = UserEmail.objects.filter(
+            email__iexact=email,
+            is_verified=True
+        ).order_by('id')
 
-class SAML2MetadataView(AuthView):
-    def dispatch(self, request, organization_slug):
-        try:
-            organization = Organization.objects.get(
-                slug=organization_slug
+        if user_emails:
+            users = User.objects.filter(
```

Comment:
Same comment here about coercing this to a list first. But it seems the difference here is that you only care about 0, 1, or more than 1 case. In which we can make this:

```python
users = list(User.objects.filter(...)[0:2])
```

This will limit the query to return at most 2 results, which satisfies what you're trying to accomplish without pulling all results.